src/data/complex/exponential.lean

```diff
@@ -1510,6 +1526,75 @@ calc cos 2 = cos (2 * 1) : congr_arg cos (mul_one _).symm
           zero_le_two) _
   ... < 0 : by norm_num
 
+lemma exp_bound_div_one_sub_of_interval_approx  {x : ℝ} (h1 : 0 ≤ x) (h2 : x ≤ 1) :
+  ∑ (j : ℕ) in finset.range 3, x ^ j / (j.factorial)
+  + x ^ 3 * ((3 : ℕ) + 1) / ((3 : ℕ).factorial * (3 : ℕ))
+  ≤ ∑ j in (finset.range 3), x ^ j :=
+begin
+  norm_num [finset.sum],
+  rw [add_assoc, add_comm (x + 1) (x ^ 3 * 4 / 18), ← add_assoc, add_le_add_iff_right,
+      ← add_le_add_iff_left (-(x^2/2)), ← add_assoc, comm_ring.add_left_neg (x^2/2), zero_add,
+      neg_add_eq_sub, sub_half, sq, pow_succ, sq],
+  have i1 : x * 4 / 18 ≤ 1 / 2 := by linarith,
+  have i2 : 0 ≤ x * 4 / 18 := by linarith,
+  have i3 := mul_le_mul h1 h1 (le_refl (0 : ℝ)) h1,
+  rw zero_mul at i3,
+  have t := mul_le_mul (le_refl (x * x)) i1 i2 i3,
+  rw ← mul_assoc,
+  rwa [mul_one_div, ← mul_div_assoc, ← mul_assoc] at t,
+end
+
+lemma exp_bound_div_one_sub_of_interval {x : ℝ} (h1 : 0 ≤ x) (h2 : x < 1) :
+  real.exp x ≤ 1 / (1 - x) :=
+begin
+  have h : ∑ j in (finset.range 3), x^j ≤ 1/(1-x),
+  { norm_num [finset.sum],
+    have h1x : 0 < 1 - x := by simpa,
+    apply (le_div_iff h1x).2,
+    norm_num [← add_assoc, mul_sub_left_distrib, mul_one, add_mul,
+      sub_add_eq_sub_sub, pow_succ' x 2],
+    have hx3 : 0 ≤ x ^ 3,
+    { norm_num,
+      exact h1 },
+    linarith },
+  exact le_trans (exp_bound' h1 h2.le (by linarith))
+   (le_trans (exp_bound_div_one_sub_of_interval_approx h1 h2.le) h),
+end
+
+lemma one_sub_le_exp_minus_of_pos {y : ℝ} (h : 0 ≤ y) : 1 - y ≤ real.exp (-y) :=
+begin
+  rw real.exp_neg,
+  have r1 : (1 - y) * (real.exp y) ≤ 1,
+  { cases le_or_lt (1-y) 0, {
+    have h'' : (1 - y) * real.exp y ≤ 0,
```

Comment:
```suggestion
  { cases le_or_lt (1 - y) 0,
    { have h'' : (1 - y) * y.exp ≤ 0,
```

i've never seen this bracket convention in any other programming language, but I quite like it tbh. we did just get a linter for this so I'm surprised it didn't shout at you about it.

Reply:
also, it's nice to note that if you want to rename the stuff that `cases` spits out, you can use `cases [...] with h₁ h₁` (this would work in this example, but I recommend toying with it generally). also, `rcases` and `obtain` are really snazzy tactics that are worth looking into - especially their interaction with the word `rfl`!